### PR TITLE
Home Task Day 3: File System & Streams

### PR DESCRIPTION
**Pull request by Masalovych**
- Generate csv file with events data (id, title, location, date, hour, etc)​
- Create GET /events?location=lviv endpoint which returns events from csv file in json format. It should support possible filtering events by location (passed as query parameter).​
- Create GET /events/:eventId endpoint for getting some specific event by id.​
- Create DELETE /events/:eventId endpoint for deleting some specific event by id.​
- Create POST /events endpoint for saving new event to the csv file.​
- Create PUT /events/:eventId endpoint for replacing specific event data in csv file.​
- Create GET /events-batch endpoint which returns all events in json format via streaming directly from csv file.​
- Optional: Make all endpoints able to properly process concurrent requests.

### CODE REVIEW

File: app/server/events/index.js [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+function convertEventsToCsv(events) {
+  return events.reduce((memo, event) => {
+    memo += `${[event.id, event.title, event.location, event.date, event.hour].join(',')}\n`;
+    return memo;
+  }, '')
+}
+
+function saveEventsToFile(events) {
+  fs.writeFileSync(dataFilePath, convertEventsToCsv(events));
```
Comment (lionvs) [inline app/server/events/index.js:35]: The issue with synchronous write is that we won't be able to process any other requests (they may fail to connect to the server) while one request is doing this write.

File: app/server/events/index.js [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+// Handlers
+
+// curl localhost:3000/events?location=lviv
+async function getEvents (req, res) {
+  const location = req.query.location;
+  getAllEvents().then(events => {
+    if (location) {
+      events = events.filter(event => event.location === location);
```
Comment (lionvs) [inline app/server/events/index.js:45]: Potentially we could avoid loading all the events into the memory by filtering needed events while reading data from the file (using streams). The node process would use less memory.

File: app/server/events/index.js [hunk 3]
```diff
@@ -0,0 +1,138 @@
+const stream = require('stream');
+const fs = require('fs');
+const path = require('path');
+const csv = require('csv-parser');
+
+const dataFilePath = path.join(__dirname, '..', '..', 'data', 'events.csv');
+const dataHeaders = ['id', 'title', 'location', 'date', 'hour'];
+
+function getAllEvents(normalized) {
```
Comment (lionvs) [inline app/server/events/index.js:9]: We could move methods that work with `csv` data to another file. We would divide responsibilities between modules/files. One file would be responsible for the `HTTP` layer and another one for `data/csv` layer.

File: app/server/events/index.js [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+async function createEvent (req, res) {
+  const {id, title, location, date, hour} = req.body;
+
+  getAllEvents().then(events => {
+    events.push({id, title, location, date, hour});
+    saveEventsToFile(events);
+
+    res.json([{result: 'Event was added'}])
+  });
```
Comment (lionvs) [inline app/server/events/index.js:96]: In the case of `error` we won't notify the client about it, the request will timeout. `catch` block could take care of that.

**Review by lionvs — APPROVED** [review]
Looks Great!

I encourage you to also think about such potential issue:

If we have concurrent requests for PUT endpoint we can lose some updates.
For example (R1- request 1, R2-request2, ):
R1 wants to update event with id 5
R2 wants to update event with id 6
R1 reads data from file
R2 reads data from file
R1 updates event with id 5 to the data
R2 updates event with id 6 to the data (note that event with id 5 is not updated because data was read before its update)
If we now query all events we could see that event with id 5 is not updated.

File: app/server/events/index.js (hunk 4 above)
Comment (Masalovych) [reply app/server/events/index.js:96]: Thanks. Will fix.

File: app/server/events/index.js (hunk 3 above)
Comment (Masalovych) [reply app/server/events/index.js:9]: Ok. Will move parser to another file.

File: app/server/events/index.js (hunk 1 above)
Comment (Masalovych) [reply app/server/events/index.js:35]: Yes. My server read from stream but write async. Will fix.

File: app/server/events/index.js (hunk 2 above)
Comment (Masalovych) [reply app/server/events/index.js:45]: Ok. Try to implement. Thanks.

**Pull request merged**